### PR TITLE
Implement LoRA finetuning for Flux.1-dev

### PR DESCRIPTION
**Pull request by mutatedducks97**
# Overview
As the title suggests, this PR implements Flux.1-dev LoRA finetuning. The Flux models are a family of text-to-image generation models released by Black Forest Labs. The model of interst in this PR is the `black-forest-labs/FLUX.1-dev` model which is 12B parameters.

The Flux Pipeline is composed of 
1. Encoder: T5, CLIP
2. Transformer (Diffuser): which estimates the noise removal. In Stable Diffusion this is carried out by a unet.
3. Noise Scheduler:  which manages the magnitude of denoising process at each interation
4. VAE: Encodes and Decodes images into a latent space 

Development and Testing was done on a 
1x A100 (40 GB SXM4)
30 vCPUs, 200 GiB RAM, 0.5 TiB SSD

# Contributions
Sample config
   - src/invoke_training/sample_configs/flux_lora_1x40gb.yaml
 

### Config
   - Job Config: src/invoke_training/pipelines/flux/lora/config.py
   - Data Loader Config: src/invoke_training/config/data/data_loader_config.py
   - Pipeline config: src/invoke_training/config/pipeline_config.py

### DataLoading
   - src/invoke_training/_shared/data/data_loaders/image_caption_flux_dataloader.py
   - src/invoke_training/_shared/data/transforms/flux_image_transform.py
        - resizes and crops images to be BxW, primarily 512x512.

### Load models from HuggingFace Pipeline:
   - src/invoke_training/_shared/flux/model_loading_utils.py
        - Models loaded from HuggingFace [FluxPipeline](https://huggingface.co/docs/diffusers/main/en/api/pipelines/flux)
        - CLIP encoder, T5 encoder, [Transformer](https://huggingface.co/docs/diffusers/main/en/api/models/flux_transformer): FluxTransformer2DModel, VAE
        - In theory we could have loaded models from [Black Forest Lab](https://github.com/black-forest-labs/flux) repo directly; however, this would likely not work with PEFT out of the box. There is a LoRA implementation but is unclear how stable and reliable the repo is. Kohya-ss does use the [BFL implementation](https://github.com/kohya-ss/sd-scripts/blob/sd3/library/flux_models.py#L878) implementation  but reworked in their repo.
      
### Utils for preprocessing text 
   - src/invoke_training/_shared/flux/encoding_utils.py

### Main training entrypoint
   - Contains main forward,src/invoke_training/pipelines/flux/lora/train.py
        - Loads data, Embeds prompt with CLIP and T5
        - Inject Lora Layer using [PEFT](https://github.com/huggingface/peft)
        - Encode image with VAE to latent, use [Flow Matching Euler Scheduling ](https://huggingface.co/docs/diffusers/en/api/schedulers/flow_match_euler_discrete) for generating timesteps and noised latents
        - forward pass of [Flux Transformer](https://huggingface.co/docs/diffusers/main/en/api/models/flux_transformer) 
        - Computes [Flow Matching Loss](https://discuss.huggingface.co/t/meaning-of-vector-fields-in-flux-and-sd3-loss-function/106601). See basic theoretical intuition here [resource](https://ronaldyu.substack.com/p/an-intuitive-intro-to-flow-matching?utm_campaign=post&utm_medium=web) or (for the more mathematically inclined) see this [resource](https://mlg.eng.cam.ac.uk/blog/2024/01/20/flow-matching.html)
        - backwards pass using HuggingFace [accelerate](https://huggingface.co/docs/accelerate/en/index)

### Utils for saving models in kohya format
   - src/invoke_training/_shared/flux/lora_checkpoint_utils.py
 
### Generating validation images
   - src/invoke_training/_shared/flux/validation.py 

### Running
Begin script with sample config using the following command
`invoke-train --cfg-file src/invoke_training/sample_configs/flux_lora_1x40gb.yaml`

### Testing:
Results with 10 (top left), 100 (bottom left), 200 (top right), 350 (bottom right) epochs of training. 
 - Training Data
image

  - Caption: A stuffed gnome at the beach with a pina colada in its hand.

image
  - Caption: A stuffed gnome reading a book in a cozy library.

image
  - Caption: A stuffed gnome sitting in a garden surrounded by colorful flowers and butterflies.

image

#### Notable features learned from sample_data:
- muting colors of outfit
- remove ears
- enlarged beard, disheveling of beard
-  nose
- increased texture on body, clothes, and hat

# ~TO-DO~ Done
### Higher Urgency:
- [x] Numerical instability while training causing massive gradients. Will need to investigate this thoroughly
- [x] Ensure Scheduler and Loss is correctly in sync

If the reviewer has any suggestions or have ran into similar issues with SD3 any tips are welcome and appreciated.

### Lower Urgency:
- [x] Clean up - removing unused code, duplicated code, and comments
- [x] Naming convention. I use `diffuser` to refer to FluxTransformer2DModel. It should be `transformer` or `flux_transformer` as to not be confused with HF transformers or diffusers models.

# Future improvements
-  Include more sampling methods as per Section 3.1: https://arxiv.org/pdf/2403.03206v1: Logit-Normal, Mode, Cosma
Default to uniform. The Theoretical intuition is more examples from middle of or middle end of the 1000 timesteps are most valuable to learn from since they are the most difficult diffusion step, so we should sample those.
- Memory optimization: See  [here](https://huggingface.co/docs/diffusers/main/en/api/pipelines/flux#optimize) and [here](https://huggingface.co/black-forest-labs/FLUX.1-schnell/discussions/5#66acc08479d0fc2e380a7f5b) for discussion on VRAM requirements and potential optimization.

# Final Tests
50, 200 epochs posted below.
image

### CODE REVIEW

File: src/invoke_training/sample_configs/flux_lora_1x40gb.yaml [hunk 1]
```diff
@@ -0,0 +1,51 @@
+# Training mode: LoRA
+# Base model:    Flux.1-dev
+# Dataset:       Bruce the Gnome
+# GPU:           1 x 8GB
```
Comment (mutatedducks97) [inline src/invoke_training/sample_configs/flux_lora_1x40gb.yaml:4]:
```suggestion
# GPU:           1 x 40GB
```

**Comment by mutatedducks97** [comment]
TODO: 

- [ ] Investigate token limit for CLIP & T5: It looks like Kohya has relaxed this to 256 or 512 for flux.1-dev [here](https://github.com/kohya-ss/sd-scripts/issues/1729
). It sounds like we can relax the [T5 encoder to 512](https://huggingface.co/black-forest-labs/FLUX.1-dev/discussions/43#6703ebb6e5a6a4fb17bbefe4 (edited) 
) at the least

File: src/invoke_training/_shared/flux/lora_checkpoint_utils.py [hunk 2]
```diff
@@ -140,3 +143,379 @@ def convert_flux_peft_checkpoint_to_kohya_state_dict(
...
+        v_B = source_dict.pop(f"{prefix}{block_prefix}attn.to_v.lora_B.weight")
+        mlp_A = source_dict.pop(f"{prefix}{block_prefix}proj_mlp.lora_A.weight")
+        mlp_B = source_dict.pop(f"{prefix}{block_prefix}proj_mlp.lora_B.weight")
+        
+        target_dict[f"single_blocks.{block_idx}.linear1.lora_A.weight"] = torch.cat(
+            [q_A, k_A, v_A, mlp_A], dim=0
+        )
+        target_dict[f"single_blocks.{block_idx}.linear1.lora_B.weight"] = torch.cat(
+            [q_B, k_B, v_B, mlp_B], dim=0
+        )
```
Comment (mutatedducks97) [inline src/invoke_training/_shared/flux/lora_checkpoint_utils.py:366]: 4x

File: src/invoke_training/_shared/flux/lora_checkpoint_utils.py [hunk 3]
```diff
@@ -140,3 +143,379 @@ def convert_flux_peft_checkpoint_to_kohya_state_dict(
...
+        target_dict[f"double_blocks.{block_idx}.img_attn.qkv.lora_B.weight"] = torch.cat(
+            [sample_q_B, sample_k_B, sample_v_B], dim=0
+        )
+        target_dict[f"double_blocks.{block_idx}.txt_attn.qkv.lora_A.weight"] = torch.cat(
+            [context_q_A, context_k_A, context_v_A], dim=0
+        )
+        target_dict[f"double_blocks.{block_idx}.txt_attn.qkv.lora_B.weight"] = torch.cat(
+            [context_q_B, context_k_B, context_v_B], dim=0
+        )
```
Comment (mutatedducks97) [inline src/invoke_training/_shared/flux/lora_checkpoint_utils.py:282]: issue here the dimension is 3xrank due to this concatenation.

**Review by hipsterusername — DISMISSED** [review]

File: src/invoke_training/_shared/flux/encoding_utils.py [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+
+    # Get shape and duplicate for multiple generations
+    _, seq_len, _ = prompt_embeds.shape
+    prompt_embeds = prompt_embeds.repeat(1, num_images_per_prompt, 1)
+    prompt_embeds = prompt_embeds.view(batch_size * num_images_per_prompt, seq_len, -1)
+
+    return prompt_embeds
+
+
+def handle_lora_scale(
```
Comment (hipsterusername) [inline src/invoke_training/_shared/flux/encoding_utils.py:99]: Have these been tested? How would they work?

(handle lora scale/reset_lora_scale)

They both have an undefined text_encoder

File: src/invoke_training/sample_configs/sdxl_textual_inversion_masks_gnome_1x24gb.yaml [hunk 5]
```diff
@@ -17,7 +17,7 @@ data_loader:
   type: TEXTUAL_INVERSION_SD_DATA_LOADER
   dataset:
     type: IMAGE_CAPTION_JSONL_DATASET
-    jsonl_path: sample_data/bruce_the_gnome/data_masks.jsonl
```
Comment (hipsterusername) [inline src/invoke_training/sample_configs/sdxl_textual_inversion_masks_gnome_1x24gb.yaml:20]: why is this changed?

File: src/invoke_training/_shared/data/transforms/flux_image_transform.py [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+            # Determine the target image resolution.
+            if self.resolution is not None:
+                resolution = self.resolution
+                resolution_obj = Resolution(resolution, resolution)
+            else:
+                original_size_hw = (image.height, image.width)
+                resolution_obj = self.aspect_ratio_bucket_manager.get_aspect_ratio_bucket(Resolution.parse(original_size_hw))
+
+            image = resize_to_cover(image, resolution_obj)
+            image = transforms.CenterCrop(resolution)(image)
```
Comment (hipsterusername) [inline src/invoke_training/_shared/data/transforms/flux_image_transform.py:57]: Noting: this hard codes center crop

File: src/invoke_training/_shared/flux/lora_checkpoint_utils.py [hunk 7]
```diff
@@ -0,0 +1,514 @@
...
+import peft
+import torch
+from diffusers import FluxTransformer2DModel
+from transformers import CLIPTextModel
+
+from invoke_training._shared.checkpoints.serialization import save_state_dict
+from invoke_training._shared.checkpoints.lora_checkpoint_utils import (
+    save_multi_model_peft_checkpoint,
+    load_multi_model_peft_checkpoint,
+    _convert_peft_models_to_kohya_state_dict,
```
Comment (hipsterusername) [inline src/invoke_training/_shared/flux/lora_checkpoint_utils.py:13]: Imported here, but defined later in the doc

File: src/invoke_training/_shared/flux/lora_checkpoint_utils.py [hunk 8]
```diff
@@ -0,0 +1,514 @@
...
+        target_dict,
+        source_dict,
+        f"{prefix}{block_prefix}norm1_context.linear.weight",
+        f"double_blocks.{block_idx}.txt_mod.lin.weight",
+    )
+
+    # Convert attention weights by concatenating Q, K, V
+    try:
+        # Sample attention weights
+        sample_q_A = source_dict.pop(f"{prefix}{block_prefix}attn.to_q.lora_A.weight")
```
Comment (hipsterusername) [inline src/invoke_training/_shared/flux/lora_checkpoint_utils.py:265]: variables failing ruff

File: src/invoke_training/pipelines/flux/lora/train.py [hunk 9]
```diff
@@ -0,0 +1,673 @@
...
+    config: FluxLoraConfig,
+    data_batch: dict,
+    vae: AutoencoderKL,
+    noise_scheduler: FlowMatchEulerDiscreteScheduler,
+    tokenizer_1: CLIPTokenizer,
+    tokenizer_2: T5Tokenizer,
+    text_encoder_1: CLIPTextModel,
+    text_encoder_2: T5EncoderModel,
+    transformer: FluxTransformer2DModel | PeftModel,
+    weight_dtype: torch.dtype,
```
Comment (hipsterusername) [inline src/invoke_training/pipelines/flux/lora/train.py:222]: seems like these three are defined but not accessed. This was mentioned in an earlier comment re: items in the config (like masks) -- Confirming, these aren't actually used and the feature isn't supported?

File: src/invoke_training/sample_configs/sdxl_textual_inversion_masks_gnome_1x24gb.yaml (hunk 5 above)
Comment (mutatedducks97) [reply src/invoke_training/sample_configs/sdxl_textual_inversion_masks_gnome_1x24gb.yaml:20]: `sample_data/bruce_the_gnome/data_masks.jsonl` didn't exist I thought it was a type-o. I am realizing that it is specific to `...masks_gnome_...yaml` in the title of these configs.

File: src/invoke_training/_shared/flux/encoding_utils.py (hunk 4 above)
Comment (mutatedducks97) [reply src/invoke_training/_shared/flux/encoding_utils.py:99]: Yeah its basically just part of the standard forward pass for the hugginface diffusers pipeline. I just pulled out these parts into a function. 

https://github.com/huggingface/diffusers/blob/main/src/diffusers/pipelines/flux/pipeline_flux.py#L374-L382
https://github.com/huggingface/diffusers/blob/main/src/diffusers/pipelines/flux/pipeline_flux.py#L346-L354

File: src/invoke_training/pipelines/flux/lora/train.py (hunk 9 above)
Comment (mutatedducks97) [reply src/invoke_training/pipelines/flux/lora/train.py:230]: we need this for encoding the prompt in line 259

File: src/invoke_training/_shared/data/transforms/flux_image_transform.py (hunk 6 above)
Comment (mutatedducks97) [reply src/invoke_training/_shared/data/transforms/flux_image_transform.py]: sd/sdxl defaults to center_crop, so i did too. I added fallback as random crop

File: src/invoke_training/_shared/flux/lora_checkpoint_utils.py (hunk 7 above)
Comment (mutatedducks97) [reply src/invoke_training/_shared/flux/lora_checkpoint_utils.py]: removed

**Review by hipsterusername — APPROVED** [review]
